File: scripts/omt/net/analysis.py
"""Analysis layer for the Petri-net library (doc §11–§24, §31; design_001 §6).

HARNESS CLONE (feature_039, D2): byte-faithful copy of
`src/agentx/model/petri_net/analysis.py` for the meta-harness net layer
(meta_harness_concurrent) — the ONLY delta is the import
(`agentx.model.petri_net.model` → `.model`, relative inside the package).
Parity pinned by the 9 shared conformance vectors. Do NOT extend here (F6).

The analysis layer is a REASONING layer, not a simulation layer (§26/§34): it
systematically explores all relevant possibilities (BFS over markings) or
reasons exactly (rational-arithmetic invariants) and reports precisely what it
proved, disproved, or could not decide (§39 — no overclaiming):

- ``True``  = proven, ``False`` = disproven, ``None`` = unknown/incomplete.
- Every exploration API takes a required keyword-only ``max_states`` (``int |
  None``; ``None`` = explicitly unlimited — F2). A truncated search returns
  ``complete=False`` and NEVER claims "deadlock-free", ``bounded=True``, or
  "unbounded" (§17/§39).
- Analysis never mutates the live marking — pure transformations over
  immutable tuple markings (§5.4, §34). Deterministic ordering everywhere
  (sorted places/transitions, BFS fixed by ``transition_order`` — §29).
- ``transition_liveness``/``is_live``/``strongly_connected_components``
  operate on a PRECOMPUTED :class:`ReachabilityGraph` (§32); an incomplete
  graph yields ``value=None, complete=False``.
- SCC note: components are computed over the graph's vertex set
  (``graph.states``); edges to targets outside the vertex set can only come
  from a truncated graph and are dangling references (§13), so they do not
  create phantom single-state components.
"""
from __future__ import annotations

from collections import deque
from dataclasses import dataclass
from fractions import Fraction
from math import gcd
from typing import TypeAlias

from .model import PetriNet

Marking: TypeAlias = tuple[int, ...]


# ---------------------------------------------------------------------------
# Result dataclasses (§31 exact — F3: no extra fields)
# ---------------------------------------------------------------------------

@dataclass(frozen=True)
class AnalysisResult:
    value: bool | None
    complete: bool
    explored_states: int
    reason: str | None = None


@dataclass(frozen=True)
class ReachabilityResult:
    markings: frozenset[Marking]
    predecessors: dict[Marking, tuple[Marking | None, str | None]]
    complete: bool
    explored_states: int


@dataclass(frozen=True)
class ReachabilityGraph:
    states: frozenset[Marking]
    edges: dict[Marking, tuple[tuple[str, Marking], ...]]
    complete: bool


@dataclass(frozen=True)
class DeadlockResult:
    deadlocks: tuple[Marking, ...]
    complete: bool
    explored_states: int
    reason: str | None = None


@dataclass(frozen=True)
class BoundResult:
    bounded: bool | None
    bounds: dict[str, int]
    complete: bool
    reason: str | None = None


# ---------------------------------------------------------------------------
# Analyzer (constructor-bound — lock checklist resolution)
# ---------------------------------------------------------------------------

class PetriNetAnalyzer:
    """Analysis toolkit over a :class:`PetriNet` (never mutates its marking)."""

    def __init__(self, net: PetriNet) -> None:
        self.net = net

    # ------------------------------------------------------------------
    # Shared BFS exploration core (§31)
    # ------------------------------------------------------------------

    def _explore(
        self, *, max_states: int | None
    ) -> tuple[
        frozenset[Marking],
        dict[Marking, tuple[Marking | None, str | None]],
        dict[Marking, tuple[tuple[str, Marking], ...]],
        bool,
        int,
    ]:
        """BFS from M0. Truncation finishes the current state's edges, then
        stops: edges to unvisited successors are recorded (§13), the
        successors are not enqueued, and ``complete`` is False.
        ``explored_states`` = distinct visited markings incl. the initial."""
        net = self.net
        initial = net.initial_marking_tuple()
        queue: deque[Marking] = deque([initial])
        visited: set[Marking] = {initial}
        predecessors: dict[Marking, tuple[Marking | None, str | None]] = {
            initial: (None, None)
        }
        edges: dict[Marking, list[tuple[str, Marking]]] = {}
        complete = True
        while queue:
            marking = queue.popleft()
            outgoing: list[tuple[str, Marking]] = []
            for transition in net.enabled_transitions_at(marking):
                successor = net.fire_marking(marking, transition)
                outgoing.append((transition, successor))
                if successor in visited:
                    continue
                if max_states is not None and len(visited) >= max_states:
                    complete = False
                    continue
                visited.add(successor)
                predecessors[successor] = (marking, transition)
                queue.append(successor)
            edges[marking] = outgoing
            if not complete:
                break
        return (
            frozenset(visited),
            predecessors,
            {m: tuple(o) for m, o in edges.items()},
            complete,
            len(visited),
        )

    # ------------------------------------------------------------------
    # Exploration APIs (F2 signatures: required keyword-only max_states)
    # ------------------------------------------------------------------

    def reachable_markings(self, *, max_states: int | None) -> ReachabilityResult:
        states, preds, _edges, complete, n = self._explore(max_states=max_states)
        return ReachabilityResult(
            markings=states, predecessors=preds, complete=complete, explored_states=n
        )

    def reachability_graph(self, *, max_states: int | None) -> ReachabilityGraph:
        states, _preds, edges, complete, _n = self._explore(max_states=max_states)
        return ReachabilityGraph(states=states, edges=edges, complete=complete)

    def deadlocks(self, *, max_states: int | None) -> DeadlockResult:
        states, _preds, _edges, complete, n = self._explore(max_states=max_states)
        deadlocks = tuple(
            sorted(m for m in states if self.net.enabled_transitions_at(m) == [])
        )
        reason = None
        if not complete:
            reason = (
                "State-space exploration was truncated; "
                "listed deadlocks are only those among explored states."
            )
        return DeadlockResult(
            deadlocks=deadlocks, complete=complete, explored_states=n, reason=reason
        )

    def bounds(self, *, max_states: int | None) -> BoundResult:
        states, _preds, _edges, complete, _n = self._explore(max_states=max_states)
        maxima = {p: 0 for p in self.net.place_order}
        for marking in states:
            for i, tokens in enumerate(marking):
                place = self.net.place_order[i]
                if tokens > maxima[place]:
                    maxima[place] = tokens
        if complete:
            return BoundResult(bounded=True, bounds=maxima, complete=True)
        return BoundResult(
            bounded=None,
            bounds=maxima,
            complete=False,
            reason="State-space exploration was truncated; boundedness is unknown.",
        )

    # ------------------------------------------------------------------
    # Graph-driven APIs (no exploration; §32)
    # ------------------------------------------------------------------

    def firing_sequence_to(
        self, result: ReachabilityResult, target: Marking
    ) -> list[str] | None:
        """Shortest (BFS edge count) firing sequence to ``target``, or None.

        ``None`` is a PROOF of unreachability only when ``result.complete``
        is True (§14); on a truncated result it means "not found among the
        explored states".
        """
        if target not in result.markings:
            return None
        sequence: list[str] = []
        current = target
        while True:
            previous, transition = result.predecessors[current]
            if previous is None:
                break
            assert transition is not None  # non-initial predecessors carry a label
            sequence.append(transition)
            current = previous
        sequence.reverse()
        return sequence

    def transition_liveness(
        self, transition: str, graph: ReachabilityGraph
    ) -> AnalysisResult:
        """Is ``transition`` live on this COMPLETE graph (§20–§21, §32)?

        A transition is live when from every reachable state an enabling
        state is reachable (reverse multi-source BFS). Incomplete graph ->
        ``value=None`` (unknown), never a bare bool.
        """
        if not graph.complete:
            return AnalysisResult(
                None, False, len(graph.states),
                "Reachability graph is incomplete; liveness is unknown.",
            )
        enabling = {
            s for s in graph.states if self.net.is_enabled_at(s, transition)
        }
        if not enabling:
            return AnalysisResult(False, True, len(graph.states))
        reverse: dict[Marking, list[Marking]] = {s: [] for s in graph.states}
        for state in sorted(graph.states):
            for _t, successor in graph.edges.get(state, ()):
                if successor in reverse:
                    reverse[successor].append(state)
        can_reach = set(enabling)
        stack = sorted(enabling)
        while stack:
            state = stack.pop()
            for predecessor in reverse.get(state, ()):
                if predecessor not in can_reach:
                    can_reach.add(predecessor)
                    stack.append(predecessor)
        return AnalysisResult(
            can_reach == set(graph.states), True, len(graph.states)
        )

    def is_live(self, graph: ReachabilityGraph) -> AnalysisResult:
        """Is the whole net live on this COMPLETE graph (every transition)?

        Empty net -> ``AnalysisResult(True, True, 1)`` (F1: §31 uniform rule,
        ``explored_states == len(graph.states)``).
        """
        if not graph.complete:
            return AnalysisResult(
                None, False, len(graph.states),
                "Reachability graph is incomplete; global liveness is unknown.",
            )
        for transition in self.net.transition_order:
            result = self.transition_liveness(transition, graph)
            if result.value is not True:
                return result
        return AnalysisResult(True, True, len(graph.states))

    def strongly_connected_components(
        self, graph: ReachabilityGraph
    ) -> list[frozenset[Marking]]:
        """Tarjan SCCs over the graph's vertex set (§23).

        Recursive Tarjan (recursion limit fine for v1 test nets). Neighbors
        are followed in edge-tuple order, start nodes in sorted-state order,
        and edge targets outside ``graph.states`` (possible only on truncated
        graphs) are not treated as vertices. Empty net ->
        ``[frozenset({()})]``.
        """
        indices: dict[Marking, int] = {}
        lowlinks: dict[Marking, int] = {}
        on_stack: set[Marking] = set()
        stack: list[Marking] = []
        components: list[frozenset[Marking]] = []
        counter = 0
        states = set(graph.states)

        def strongconnect(v: Marking) -> None:
            nonlocal counter
            indices[v] = lowlinks[v] = counter
            counter += 1
            stack.append(v)
            on_stack.add(v)
            for _t, w in graph.edges.get(v, ()):
                if w not in states:
                    continue
                if w not in indices:
                    strongconnect(w)
                    lowlinks[v] = min(lowlinks[v], lowlinks[w])
                elif w in on_stack:
                    lowlinks[v] = min(lowlinks[v], indices[w])
            if lowlinks[v] == indices[v]:
                component: set[Marking] = set()
                while True:
                    w = stack.pop()
                    on_stack.discard(w)
                    component.add(w)
                    if w == v:
                        break
                components.append(frozenset(component))

        for v in sorted(graph.states):
            if v not in indices:
                strongconnect(v)
        return components

    # ------------------------------------------------------------------
    # Exact algebra (§7/§18/§19; D4 zero-dependency; F6/F7)
    # ------------------------------------------------------------------

    def incidence_matrix(self) -> list[list[int]]:
        """``C[p][t] = W(t,p) - W(p,t)``; rows=``place_order``, cols=``transition_order``."""
        return [
            [
                self.net.outputs[t].get(p, 0) - self.net.inputs[t].get(p, 0)
                for t in self.net.transition_order
            ]
            for p in self.net.place_order
        ]

    def place_invariants(self) -> list[tuple[int, ...]]:
        """Basis of ``Cᵀ x = 0`` (token-conservation laws), coprime int tuples.

        Degenerate nets (F7): places-but-no-transitions -> identity basis;
        empty net -> ``[]``.
        """
        matrix = self.incidence_matrix()
        n_places = len(self.net.place_order)
        if n_places == 0:
            return []
        transposed = [list(row) for row in zip(*matrix)] if matrix else []
        return [
            _coprime_int_vector(v)
            for v in nullspace(transposed, n_cols=n_places)
        ]

    def transition_invariants(self) -> list[tuple[int, ...]]:
        """Basis of ``C y = 0`` (cyclic firing multisets), coprime int tuples.

        Doc caveat: a non-negative T-invariant is a necessary but not always
        realizable firing count vector (§19). Degenerate nets (F7):
        transitions-but-no-places -> identity basis; empty net -> ``[]``.
        """
        matrix = self.incidence_matrix()
        n_trans = len(self.net.transition_order)
        if n_trans == 0:
            return []
        return [
            _coprime_int_vector(v)
            for v in nullspace(matrix, n_cols=n_trans)
        ]


# ---------------------------------------------------------------------------
# Exact rational nullspace (§18 — D4: pure-Python, zero dependencies)
# ---------------------------------------------------------------------------

def nullspace(
    matrix: list[list[int]], n_cols: int | None = None
) -> list[list[Fraction]]:
    """Exact nullspace basis via Fraction Gauss–Jordan to FULL RREF.

    Free columns each emit one basis vector: 1 at the free column,
    ``-rows[pivot_row[c]][f]`` at each pivot column ``c``, 0 elsewhere.
    ``n_cols`` is inferred from the first row when None (0 for an empty
    matrix) — PASS IT EXPLICITLY on degenerate (zero-row) shapes (F7).
    """
    rows = [list(map(Fraction, row)) for row in matrix]
    n_rows = len(rows)
    if n_cols is None:
        n_cols = len(rows[0]) if rows else 0
    pivot_row_of_col: dict[int, int] = {}
    r = 0
    for c in range(n_cols):
        pivot = None
        for i in range(r, n_rows):
            if rows[i][c] != 0:
                pivot = i
                break
        if pivot is None:
            continue
        rows[r], rows[pivot] = rows[pivot], rows[r]
        factor = rows[r][c]
        rows[r] = [x / factor for x in rows[r]]
        for i in range(n_rows):
            if i != r and rows[i][c] != 0:
                f = rows[i][c]
                rows[i] = [x - f * y for x, y in zip(rows[i], rows[r])]
        pivot_row_of_col[c] = r
        r += 1
        if r == n_rows:
            break
    basis: list[list[Fraction]] = []
    for f in range(n_cols):
        if f in pivot_row_of_col:
            continue
        vec = [Fraction(0)] * n_cols
        vec[f] = Fraction(1)
        for c, pivot_row in pivot_row_of_col.items():
            vec[c] = -rows[pivot_row][f]
        basis.append(vec)
    return basis


def _coprime_int_vector(vec: list[Fraction]) -> tuple[int, ...]:
    """Deterministic integer representative: LCM-scale to ints, divide by the
    gcd-content, negate when the first nonzero component is negative (§19)."""
    lcm = 1
    for x in vec:
        lcm = lcm * x.denominator // gcd(lcm, x.denominator)
    ints = [int(x * lcm) for x in vec]
    content = 0
    for v in ints:
        content = gcd(content, abs(v))
    if content > 1:
        ints = [v // content for v in ints]
    for v in ints:
        if v != 0:
            if v < 0:
                ints = [-x for x in ints]
            break
    return tuple(ints)

File: tools/petri-net-studio/src/engine/analysis.ts
/**
 * Analysis layer — TS port of `src/agentx/model/petri_net/analysis.py`
 * (executable spec; 38 behaviors in `tests/engine/analysis.test.ts`).
 * Semantics pinned by design_001 §5: exact parity B2/B3, deterministic
 * ordering B6, no overclaiming (truncation → value null/complete false).
 *
 * Port notes:
 * - B1: markings keyed by `markingKey(m) = m.join(",")` (non-negative ints
 *   ⇒ unambiguous). Results re-materialize arrays from keys in deterministic
 *   (compareMarkings-sorted) order.
 * - B2: `nullspace` copies analysis.py line-for-line (Fraction Gauss–Jordan
 *   to FULL RREF); `_coprimeIntVector` mirrors `_coprime_int_vector`
 *   (LCM-scale → gcd-content divide → negate if first nonzero negative).
 *   No floats in the algebra path (fraction.ts).
 * - B3: `explore` copies the Python `_explore` loop 1:1 — edge-recording +
 *   no-enqueue + `complete=false` + finish-current-state-edges + break.
 * - B7: Tarjan skips edge targets outside `graph.states` (truncation-only
 *   dangling references — no phantom single-state components).
 */

import { PetriNet, Marking } from "./model.js";
import { Fraction, gcd, lcm } from "./fraction.js";

/** B1 — markings are non-negative ints ⇒ join(",") is unambiguous. */
export function markingKey(m: Marking): string {
  return m.join(",");
}

/** Key → marking (complement of markingKey; empty marking ⇒ []). */
export function markingFromKey(key: string): number[] {
// TA: gotcha: gotcha: markingFromKey must special-case the empty marking — markingKey([]) = "" and "".split(",") would yield [""] → [0] (wrong). The empty net's single SCC component must be [[]] not [[0]].
  return key === "" ? [] : key.split(",").map(Number);
}

/** Numeric lexicographic — Python tuple ordering (B6). */
export function compareMarkings(a: Marking, b: Marking): number {
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) {
    if (a[i] !== b[i]) return a[i] - b[i];
  }
  return a.length - b.length;
}

export interface AnalysisResult {
  value: boolean | null;
  complete: boolean;
  exploredStates: number;
  reason?: string | null;
}

export interface ReachabilityResult {
  markings: number[][];
  predecessors: Map<string, { prev: number[] | null; transition: string | null }>;
  complete: boolean;
  exploredStates: number;
}

export interface ReachabilityGraph {
  states: number[][];
  edges: Map<string, Array<[string, number[]]>>;
  complete: boolean;
}

export interface DeadlockResult {
  deadlocks: number[][];
  complete: boolean;
  exploredStates: number;
  reason?: string | null;
}

export interface BoundResult {
  bounded: boolean | null;
  bounds: Array<[string, number]>;
  complete: boolean;
  reason?: string | null;
}

interface ExploreOut {
  visitedMarkings: number[][];
  predecessors: Map<string, { prev: number[] | null; transition: string | null }>;
  edges: Map<string, Array<[string, number[]]>>;
  complete: boolean;
  exploredStates: number;
}

const TRUNCATED_DEADLOCK_REASON =
  "State-space exploration was truncated; " +
  "listed deadlocks are only those among explored states.";
const TRUNCATED_BOUNDS_REASON =
  "State-space exploration was truncated; boundedness is unknown.";
const INCOMPLETE_LIVENESS_REASON =
  "Reachability graph is incomplete; liveness is unknown.";
const INCOMPLETE_GLOBAL_LIVENESS_REASON =
  "Reachability graph is incomplete; global liveness is unknown.";

export class PetriNetAnalyzer {
  constructor(readonly net: PetriNet) {}

  // ------------------------------------------------------------------
  // Shared BFS exploration core (§31 — B3: 1:1 port of Python _explore)
  // ------------------------------------------------------------------

  private explore(maxStates: number | null): ExploreOut {
    const net = this.net;
    const initial = net.initialMarkingTuple();
    const queue: number[][] = [initial];
    const visited = new Map<string, number[]>();
    const initialKey = markingKey(initial);
    visited.set(initialKey, initial);
    const predecessors = new Map<
      string,
      { prev: number[] | null; transition: string | null }
    >();
    predecessors.set(initialKey, { prev: null, transition: null });
    const edges = new Map<string, Array<[string, number[]]>>();
    let complete = true;
    let qi = 0;
    while (qi < queue.length) {
      const marking = queue[qi++];
      const outgoing: Array<[string, number[]]> = [];
      for (const transition of net.enabledTransitionsAt(marking)) {
        const successor = net.fireMarking(marking, transition);
        outgoing.push([transition, successor]);
        const skey = markingKey(successor);
        if (visited.has(skey)) {
          continue;
        }
        if (maxStates !== null && visited.size >= maxStates) {
          complete = false;
          continue;
        }
        visited.set(skey, successor);
        predecessors.set(skey, { prev: marking, transition });
        queue.push(successor);
      }
      edges.set(markingKey(marking), outgoing);
      if (!complete) {
        break;
      }
    }
    return {
      visitedMarkings: [...visited.values()],
      predecessors,
      edges,
      complete,
      exploredStates: visited.size,
    };
  }

  // ------------------------------------------------------------------
  // Exploration APIs (B6: markings/states sorted by compareMarkings)
  // ------------------------------------------------------------------

  reachableMarkings(maxStates: number | null): ReachabilityResult {
    const { visitedMarkings, predecessors, complete, exploredStates } =
      this.explore(maxStates);
    return {
      markings: [...visitedMarkings].sort(compareMarkings),
      predecessors,
      complete,
      exploredStates,
    };
  }

  reachabilityGraph(maxStates: number | null): ReachabilityGraph {
    const { visitedMarkings, edges, complete } = this.explore(maxStates);
    return {
      states: [...visitedMarkings].sort(compareMarkings),
      edges,
      complete,
    };
  }

  deadlocks(maxStates: number | null): DeadlockResult {
    const { visitedMarkings, complete, exploredStates } = this.explore(maxStates);
    const deadlocks = visitedMarkings
      .filter((m) => this.net.enabledTransitionsAt(m).length === 0)
      .sort(compareMarkings);
    return {
      deadlocks,
      complete,
      exploredStates,
      reason: complete ? null : TRUNCATED_DEADLOCK_REASON,
    };
  }

  bounds(maxStates: number | null): BoundResult {
    const { visitedMarkings, complete } = this.explore(maxStates);
    const placeOrder = this.net.placeOrder;
    const maxima = new Map<string, number>(placeOrder.map((p) => [p, 0]));
    for (const marking of visitedMarkings) {
      for (let i = 0; i < marking.length; i++) {
        const place = placeOrder[i];
        if (marking[i] > (maxima.get(place) ?? 0)) {
          maxima.set(place, marking[i]);
        }
      }
    }
    const bounds = placeOrder.map((p) => [p, maxima.get(p)!] as [string, number]);
    if (complete) {
      return { bounded: true, bounds, complete: true, reason: null };
    }
    return {
      bounded: null,
      bounds,
      complete: false,
      reason: TRUNCATED_BOUNDS_REASON,
    };
  }

  // ------------------------------------------------------------------
  // Graph-driven APIs (no exploration; §32)
  // ------------------------------------------------------------------

  firingSequenceTo(result: ReachabilityResult, target: number[]): string[] | null {
    const tkey = markingKey(target);
    // null is a PROOF of unreachability only when result.complete is true.
    if (!result.markings.some((m) => markingKey(m) === tkey)) {
      return null;
    }
    const sequence: string[] = [];
    let current = target;
    while (true) {
      const entry = result.predecessors.get(markingKey(current))!;
      const { prev, transition } = entry;
      if (prev === null) {
        break;
      }
      sequence.push(transition!);
      current = prev;
    }
    sequence.reverse();
    return sequence;
  }

  transitionLiveness(transition: string, graph: ReachabilityGraph): AnalysisResult {
    // Incomplete graph -> value null (unknown), never a bare bool.
    if (!graph.complete) {
      return {
        value: null,
        complete: false,
        exploredStates: graph.states.length,
        reason: INCOMPLETE_LIVENESS_REASON,
      };
    }
    const enabling = graph.states.filter((s) => this.net.isEnabledAt(s, transition));
    if (enabling.length === 0) {
      return {
        value: false,
        complete: true,
        exploredStates: graph.states.length,
        reason: null,
      };
    }
    const reverse = new Map<string, number[][]>(
      graph.states.map((s) => [markingKey(s), []]),
    );
    for (const state of [...graph.states].sort(compareMarkings)) {
      for (const [, successor] of graph.edges.get(markingKey(state)) ?? []) {
        const skey = markingKey(successor);
        if (reverse.has(skey)) {
          reverse.get(skey)!.push(state);
        }
      }
    }
    const canReach = new Set<string>(enabling.map(markingKey));
    const stack = [...enabling].sort(compareMarkings).map(markingKey);
    while (stack.length > 0) {
      const state = stack.pop()!;
      for (const predecessor of reverse.get(state) ?? []) {
        const pkey = markingKey(predecessor);
        if (!canReach.has(pkey)) {
          canReach.add(pkey);
          stack.push(pkey);
        }
      }
    }
    return {
      value: canReach.size === graph.states.length,
      complete: true,
      exploredStates: graph.states.length,
      reason: null,
    };
  }

  isLive(graph: ReachabilityGraph): AnalysisResult {
    if (!graph.complete) {
      return {
        value: null,
        complete: false,
        exploredStates: graph.states.length,
        reason: INCOMPLETE_GLOBAL_LIVENESS_REASON,
      };
    }
    for (const transition of this.net.transitionOrder) {
      const result = this.transitionLiveness(transition, graph);
      if (result.value !== true) {
        return result;
      }
    }
    // Empty net: no transitions ⇒ AnalysisResult(True, True, 1) (F1).
    return {
      value: true,
      complete: true,
      exploredStates: graph.states.length,
      reason: null,
    };
  }

  stronglyConnectedComponents(graph: ReachabilityGraph): number[][][] {
    // Tarjan SCCs over the graph's vertex set (§23, B7). Recursive (depth
    // fine for v1 nets). Neighbors followed in edge-tuple order, start nodes
    // in sorted-state order; edge targets outside graph.states are skipped.
    const indices = new Map<string, number>();
    const lowlinks = new Map<string, number>();
    const onStack = new Set<string>();
    const stack: string[] = [];
    const components: number[][][] = [];
    let counter = 0;
    const states = new Set<string>(graph.states.map(markingKey));

    const strongconnect = (v: number[]): void => {
      const vkey = markingKey(v);
      indices.set(vkey, counter);
      lowlinks.set(vkey, counter);
      counter += 1;
      stack.push(vkey);
      onStack.add(vkey);
      for (const [, w] of graph.edges.get(vkey) ?? []) {
        const wkey = markingKey(w);
        if (!states.has(wkey)) {
          continue;
        }
        if (!indices.has(wkey)) {
          strongconnect(w);
          lowlinks.set(vkey, Math.min(lowlinks.get(vkey)!, lowlinks.get(wkey)!));
        } else if (onStack.has(wkey)) {
          lowlinks.set(vkey, Math.min(lowlinks.get(vkey)!, indices.get(wkey)!));
        }
      }
      if (lowlinks.get(vkey) === indices.get(vkey)) {
        const component: number[][] = [];
        while (true) {
          const wkey = stack.pop()!;
          onStack.delete(wkey);
          component.push(markingFromKey(wkey));
          if (wkey === vkey) {
            break;
          }
        }
        component.sort(compareMarkings);
        components.push(component);
      }
    };

    for (const v of [...graph.states].sort(compareMarkings)) {
      if (!indices.has(markingKey(v))) {
        strongconnect(v);
      }
    }
    return components;
  }

  // ------------------------------------------------------------------
  // Exact algebra (§7/§18/§19; D4 zero-dependency; F6/F7)
  // ------------------------------------------------------------------

  incidenceMatrix(): number[][] {
    // C[p][t] = W(t,p) - W(p,t); rows=placeOrder, cols=transitionOrder.
    const rows: number[][] = [];
    for (const p of this.net.placeOrder) {
      const row: number[] = [];
      for (const t of this.net.transitionOrder) {
        row.push(
          (this.net.outputs.get(t)?.get(p) ?? 0) -
            (this.net.inputs.get(t)?.get(p) ?? 0),
        );
      }
      rows.push(row);
    }
    return rows;
  }

  placeInvariants(): number[][] {
    // Basis of Cᵀ x = 0 (token-conservation laws), coprime int tuples.
    // Degenerate nets (F7): places-but-no-transitions -> identity basis;
    // empty net -> [].
    const matrix = this.incidenceMatrix();
    const nPlaces = this.net.placeOrder.length;
    if (nPlaces === 0) {
      return [];
    }
    const transposed = matrix.length > 0 ? transpose(matrix) : [];
    return nullspace(transposed, nPlaces).map(_coprimeIntVector);
  }

  transitionInvariants(): number[][] {
    // Basis of C y = 0 (cyclic firing multisets), coprime int tuples.
    // Degenerate nets (F7): transitions-but-no-places -> identity basis;
    // empty net -> [].
    const matrix = this.incidenceMatrix();
    const nTrans = this.net.transitionOrder.length;
    if (nTrans === 0) {
      return [];
    }
    return nullspace(matrix, nTrans).map(_coprimeIntVector);
  }
}

// ---------------------------------------------------------------------------
// Exact rational nullspace (§18 — D4: pure-TS, zero dependencies; B2)
// ---------------------------------------------------------------------------

function transpose(matrix: number[][]): number[][] {
  const cols = matrix.length > 0 ? matrix[0].length : 0;
  const result: number[][] = [];
  for (let c = 0; c < cols; c++) {
    const row: number[] = [];
    for (const r of matrix) {
      row.push(r[c]);
    }
    result.push(row);
  }
  return result;
}

function nullspace(matrix: number[][], nCols?: number): Fraction[][] {
  // Exact nullspace basis via Fraction Gauss–Jordan to FULL RREF (1:1 port
  // of analysis.py). Free columns each emit one basis vector: 1 at the free
  // column, -rows[pivot_row[c]][f] at each pivot column c, 0 elsewhere.
  const rows = matrix.map((row) => row.map((x) => new Fraction(x)));
  const nRows = rows.length;
  let nc = nCols;
  if (nc === undefined) {
    nc = rows.length > 0 ? rows[0].length : 0;
  }
  const pivotRowOfCol = new Map<number, number>();
  let r = 0;
  for (let c = 0; c < nc; c++) {
    let pivot: number | null = null;
    for (let i = r; i < nRows; i++) {
      if (!rows[i][c].isZero()) {
        pivot = i;
        break;
      }
    }
    if (pivot === null) {
      continue;
    }
    [rows[r], rows[pivot]] = [rows[pivot], rows[r]];
    const factor = rows[r][c];
    rows[r] = rows[r].map((x) => x.div(factor));
    for (let i = 0; i < nRows; i++) {
      if (i !== r && !rows[i][c].isZero()) {
        const f = rows[i][c];
        rows[i] = rows[i].map((x, j) => x.sub(f.mul(rows[r][j])));
      }
    }
    pivotRowOfCol.set(c, r);
    r += 1;
    if (r === nRows) {
      break;
    }
  }
  const basis: Fraction[][] = [];
  for (let f = 0; f < nc; f++) {
    if (pivotRowOfCol.has(f)) {
      continue;
    }
    const vec: Fraction[] = new Array(nc).fill(Fraction.zero());
    vec[f] = new Fraction(1);
    for (const [c, pivotRow] of pivotRowOfCol) {
      vec[c] = rows[pivotRow][f].neg();
    }
    basis.push(vec);
  }
  return basis;
}

function _coprimeIntVector(vec: Fraction[]): number[] {
// TA: gotcha: gotcha: JS -x of 0 yields -0, but Python int() canonicalizes -0.0 → 0. Vitest toEqual distinguishes 0/-0 (Object.is). Fixed in Fraction constructor AND _coprimeIntVector final map (v === 0 ? 0 : v). Caught by conformance vector weighted_reaction place_invariants ([1,-0,-2] vs [1,0,-2]).
  // Deterministic integer representative: LCM-scale to ints, divide by the
  // gcd-content, negate when the first nonzero component is negative (§19).
  let l = 1;
  for (const x of vec) {
    l = lcm(l, x.den);
  }
  const ints = vec.map((x) => x.num * (l / x.den));
  let content = 0;
  for (const v of ints) {
    content = gcd(content, Math.abs(v));
  }
  if (content > 1) {
    for (let i = 0; i < ints.length; i++) {
      ints[i] = ints[i] / content;
    }
  }
  for (const v of ints) {
    if (v !== 0) {
      if (v < 0) {
        for (let i = 0; i < ints.length; i++) {
          ints[i] = -ints[i];
        }
      }
      break;
    }
  }
  // Python int canonicalization: -0 normalizes to 0 (JS -x of 0 is -0).
  return ints.map((v) => (v === 0 ? 0 : v));
}